index.ts

```diff
@@ -468,5 +500,82 @@ export class HttpClient {
         retryNumber = Math.min(ExponentialBackoffCeiling, retryNumber);
         const ms: number = ExponentialBackoffTimeSlice*Math.pow(2, retryNumber);
         return new Promise(resolve => setTimeout(()=>resolve(), ms));
-    } 
+    }
+
+    private static dateTimeDeserializer(key: any, value: any): any {
+        if (typeof value === 'string'){
+            let a = new Date(value);
+            if (!isNaN(a.valueOf())) {
+                return a;
+            }
+        }
+
+        return value;
+    }
+
+    private async _processResponse<T>(res: ifm.IHttpClientResponse, options: ifm.IRequestOptions): Promise<ITypedResponse<T>> {
+        return new Promise<ITypedResponse<T>>(async (resolve, reject) => {
+            const statusCode: number = res.message.statusCode;
+
+            const response: ITypedResponse<T> = {
+                statusCode: statusCode,
+                result: null,
+                headers: {}
+            };
+
+            // not found leads to null obj returned
+            if (statusCode == HttpCodes.NotFound) {
+                resolve(response);
+            }
+
+            let obj: any;
+            let contents: string;
+
+            // get the result from the body
+            try {
+                contents = await res.readBody();
+                if (contents && contents.length > 0) {
+                    if (options && options.deserializeDates) {
+                        obj = JSON.parse(contents, HttpClient.dateTimeDeserializer);
+                    } else {
+                        obj = JSON.parse(contents);
+                    }
+
+                    response.result = obj;
+                }
+                
+                response.headers = res.message.headers;
+            }
+            catch (err) {
+                // Invalid resource (contents not json);  leaving result obj null
+            }
+
+            // note that 3xx redirects are handled by the http layer.
```

Comment:
> 3xx redirects are handled by the http layer

this comment is not true depending on redirect setting `IRequestOptions.maxRedirects`

Reply:
Also 304 is not an error.

Would it make more sense to change the below line to `if (statusCode > 399) {` since only 400 and above are errors?

Reply:
Nevermind, the way you have it makes more sense. This function is called when the user is expecting a payload (so generally 2xx range).

Since you added response to error, client can catch and handle appropriately.